unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/AssetManager/Material/AssetPromise_Material.cs

```diff
@@ -1,17 +1,187 @@
+                    material.SetFloat(ShaderUtils.AlphaClip, 0);
+                    break;
+                case MaterialModel.TransparencyMode.ALPHA_TEST: // ALPHATEST
+                    material.EnableKeyword("_ALPHATEST_ON");
+
+                    material.SetInt(ShaderUtils.SrcBlend, (int) UnityEngine.Rendering.BlendMode.One);
+                    material.SetInt(ShaderUtils.DstBlend, (int) UnityEngine.Rendering.BlendMode.Zero);
+                    material.SetInt(ShaderUtils.ZWrite, 1);
+                    material.SetFloat(ShaderUtils.AlphaClip, 1);
+                    material.SetFloat(ShaderUtils.Cutoff, model.alphaTest);
+                    material.SetInt("_Surface", 0);
+                    material.renderQueue = (int) UnityEngine.Rendering.RenderQueue.AlphaTest;
+                    break;
+                case MaterialModel.TransparencyMode.ALPHA_BLEND: // ALPHABLEND
+                    material.EnableKeyword("_ALPHABLEND_ON");
+
+                    material.SetInt(ShaderUtils.SrcBlend, (int) UnityEngine.Rendering.BlendMode.SrcAlpha);
+                    material.SetInt(ShaderUtils.DstBlend, (int) UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+                    material.SetInt(ShaderUtils.ZWrite, 0);
+                    material.SetFloat(ShaderUtils.AlphaClip, 0);
+                    material.renderQueue = (int) UnityEngine.Rendering.RenderQueue.Transparent;
+                    material.SetInt("_Surface", 1);
+                    break;
+                case MaterialModel.TransparencyMode.ALPHA_TEST_AND_BLEND:
+                    material.EnableKeyword("_ALPHAPREMULTIPLY_ON");
+
+                    material.SetInt(ShaderUtils.SrcBlend, (int) UnityEngine.Rendering.BlendMode.One);
+                    material.SetInt(ShaderUtils.DstBlend, (int) UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+                    material.SetInt(ShaderUtils.ZWrite, 0);
+                    material.SetFloat(ShaderUtils.AlphaClip, 1);
+                    material.renderQueue = (int) UnityEngine.Rendering.RenderQueue.Transparent;
+                    material.SetInt("_Surface", 1);
+                    break;
+            }
+        }
+
+        private AssetPromise_DCLTexture AssignTextureToMaterial(Material material, int materialProperty, TextureModel model, Action<Exception> onFail)
+        {
+            AssetPromise_DCLTexture promiseDclTexture = new AssetPromise_DCLTexture(model);
+            if (model != null)
+            {
+                promiseDclTexture.OnSuccessEvent += (x) =>
+                {
+                    SetMaterialTexture(material, materialProperty, x.texture2D);
+                };
+                promiseDclTexture.OnFailEvent += (x, error) =>
+                {
+                    onFail?.Invoke(error);
+                };
+
+                AssetPromiseKeeper_DCLTexture.i.Keep(emissionAssetPromiseDclTexture);
```

Comment:
if we are "Keeping" textures here... shouldn't we "Forget" them when Material promise is forgotten also?

Reply:
This is only called on the `OnLoad `function. If it is already loading, the `OnCancelLoading `will clean it up, and if it is already loaded the `Cleanup `method will take care of them.

I may be wrong but `Forget `is when the load didn't started yet right?

Reply:
for every `Keep` there must be a `Forget` in any scenario.
here it looks like they will only be forgotten if the load is cancelled, but what about after everything is loaded? when are texture forgotten?

Reply:
This is done in the `CleanPromises()` in the override of the `Cleanup`

![image](https://user-images.githubusercontent.com/54064092/165473968-8451b94c-5450-4d56-8758-702216a03445.png)